group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/ContextStateMachine.java

```diff
@@ -0,0 +1,176 @@
+import org.slf4j.Logger;
+
+/**
+ * ContextStateMachine is a wrapper on top of the coordinator state machine. Generally, only a single thread at a
+ * time will access this object but multiple threads may access while loading the __consumer_offsets topic partition.
+ */
+class ContextStateMachine<S extends CoordinatorShard<U>, U> implements CoordinatorPlayback<U> {
+    /**
+     * The logger.
+     */
+    private final Logger log;
+    /**
+     * The actual state machine.
+     */
+    private S coordinator;
+
+    /**
+     * The snapshot registry backing the coordinator.
+     */
+    private final SnapshotRegistry snapshotRegistry;
+
+    /**
+     * The topic partition.
+     */
+    private final TopicPartition tp;
+
+    /**
+     * The last offset written to the partition.
+     */
+    private long lastWrittenOffset;
+
+    /**
+     * The last offset committed. This represents the high
+     * watermark of the partition.
+     */
+    private long lastCommittedOffset;
+
+    ContextStateMachine(
+        LogContext logContext,
+        SnapshotRegistry snapshotRegistry,
+        S coordinator,
+        TopicPartition tp
+    ) {
+        this.log = logContext.logger(ContextStateMachine.class);
+        this.coordinator = coordinator;
+        this.snapshotRegistry = snapshotRegistry;
+        this.tp = tp;
+        this.lastWrittenOffset = 0;
+        this.lastCommittedOffset = 0;
+        snapshotRegistry.getOrCreateSnapshot(0);
+        snapshotRegistry.deleteSnapshotsUpTo(0);
+    }
+
+    /**
+     * Reverts the last written offset. This also reverts the snapshot
+     * registry to this offset. All the changes applied after the offset
+     * are lost.
+     *
+     * @param offset The offset to revert to.
+     */
+    synchronized void revertLastWrittenOffset(
+        long offset
+    ) {
+        if (offset > lastWrittenOffset) {
+            throw new IllegalStateException("New offset " + offset + " of " + tp +
+                " must be smaller than " + lastWrittenOffset + ".");
+        }
+
+        log.debug("Revert last written offset of {} to {}.", tp, offset);
+        lastWrittenOffset = offset;
+        snapshotRegistry.revertToSnapshot(offset);
+    }
+
+    @Override
+    public synchronized void replay(
+        long producerId,
+        short producerEpoch,
+        U record
+    ) {
+        coordinator.replay(producerId, producerEpoch, record);
+    }
+
+    /**
+     * Updates the last written offset. This also create a new snapshot
+     * in the snapshot registry.
+     *
+     * @param offset The new last written offset.
+     */
+    @Override
+    public synchronized void updateLastWrittenOffset(Long offset) {
+        if (offset <= lastWrittenOffset) {
+            throw new IllegalStateException("New last written offset " + offset + " of " + tp +
+                " must be greater than " + lastWrittenOffset + ".");
+        }
+
+        log.debug("Update last written offset of {} to {}.", tp, offset);
```

Comment:
nit: I have noticed that we log before updating here while we log after updating in `updateLastCommittedOffset`. It would be good to be consistent. If you put them after, I would also use `Updated` to signal that it was done.